### DISCUSSION TITLE
Is there a way to specify a target widget in a binding tuple?

### DISCUSSION THREAD

**Discussion by josephcourtney** [Textualize/textual#2207] [Q&A] [answered]
I searched past issues and didn't find anything that would allow this. The idea is that you could include a selector string in the binding tuple that would allow you to define bindings for widgets that are not stored as properties of the App object.

One way this could work is by including a css selector in the binding tuple. If it was included in the "action" element of the binding tuple:
```
class TextLogApp(App):
    BINDINGS=[
        ("j", "#content.scroll_down", "Scroll content down"),
        ("k", "#content.scroll_up", "Scroll content up"),
    ]
    def __init__(self):
        self._content = '\n'.join(f'test {i}' for i in range(200))
        super().__init__()

    def compose(self) -> ComposeResult:
        yield TextLog(id='content')

    def on_ready(self) -> None:
        text_log = self.query_one(TextLog)
        text_log.write(self._content)
```

would be equivalent to:

```
class TextLogApp(App):
    BINDINGS=[
        ("j", "content_scroll_down", "Scroll content down"),
        ("k", "content_scroll_up", "Scroll content up"),
    ]
    def __init__(self):
        self._content = '\n'.join(f'test {i}' for i in range(200))
        super().__init__()

    def compose(self) -> ComposeResult:
        yield TextLog(id='content')

    def on_ready(self) -> None:
        text_log = self.query_one(TextLog)
        text_log.write(self._content)

    def action_content_scroll_down(self):
        self.query_one('#content').scroll_down()

    def action_content_scroll_up(self):
        self.query_one('#content').scroll_up()
```

I think this could be implemented by changing line 2068 in src/textual/app.py from

`raise ActionError(f"Action namespace {destination} is not known")`

to

```
try:
    action_target = self.query_one(destination)
except NoMatches as error:
    raise error
except:
    raise ActionError(f"Action namespace {destination} is not known")
```

Another option would be the inclusion of an optional fourth element of the binding tuple containing the css selector.

I don't know if some other changes would be necessary. I apologize if there is some other intended method of achieving the same results.

**Comment by davep** [comment, marked as answer]
As you've already identified, no, there is no support for that in the binding facility as it stands. I suppose my first question is: what exactly is it you're wanting to do that would require such support? In the example you've given above it looks like you want to add vi-like navigation to `TextLog`, is that the case?

Normally, if I want to extend the bindings for a particular widget I inherit from it and add my own bindings. By way of illustration, someone asked a wee back how they could do similar for the `MarkdownViewer` and I gave this as an example:

```python
from pathlib import Path

from textual.app     import App, ComposeResult
from textual.widgets import Header, Footer, MarkdownViewer
from textual.binding import Binding

class ViLikeMarkdownViewer( MarkdownViewer ):

    BINDINGS = [
        Binding( "j", "up", "", show=False ),
        Binding( "k", "down", "", show=False ),
    ]

    def action_up( self ) -> None:
        self.scroll_up()

    def action_down( self ) -> None:
        self.scroll_down()

class ViLikeMarkdownApp( App[ None ] ):

    def compose( self ) -> ComposeResult:
        yield Header()
        yield ViLikeMarkdownViewer( Path( "../textual/README.md" ).read_text() )
        yield Footer()

if __name__ == "__main__":
    ViLikeMarkdownApp().run()
```

So you should be able to make a `ViLikeTextLog` and then use it as you'd normally use a `TextLog`.

Longer-term, on our roadmap, we do have a plan to make bindings more configurable such that different binding "themes" could be used.
